### PR TITLE
SQL auditcare

### PR DESCRIPTION
**Pull request by millerdev**
This PR converts all auditcare reads and writes to use SQL instead of Couch. There is one view that is probably unused that still reads from Couch, but that view probably has not been used in years given what it does (it will be removed in a subsequent PR).

Once this is merged there will be no user-facing way to query auditcare data stored in Couch. In the interest of keeping this PR as small as possible and to get it out as soon as possible it seemed best to drop that feature temporarily and implement it in a follow-up PR.

It will probably easiest to review by commit. :fish: 

Pre-merge checklist:
- [x] Deploy to staging
- [x] Create prod auditcare database
- [x] Create india auditcare database
- [x] Create change log for external partners

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

Automated tests have been added for all areas that were refactored in a significant way. The test coverage for auditcare is much better than it was previously.

### QA Plan

I will be testing this on staging. No other QA is needed.

### Rollback instructions

This PR contains a SQL migration that creates new tables for auditcare. If a rollback becomes necessary a decision will need to be made at that time whether to copy accumulated SQL data into Couch, or to delete it, or to do something else.

### CODE REVIEW

File: corehq/util/models.py
Comment (dannyroberts) [inline corehq/util/models.py:255]: Maybe this is such a common pattern that it would be obvious to many people, but I think if I hadn't already had the conversation with you about it, I wouldn't know that a primary use case of this would be to save space when there's a fairly small set of values but each value is somewhat long. Might be worth just spelling that out in this docstring for anyone who might not already be familiar with that pattern

File: corehq/apps/auditcare/management/commands/gdpr_scrub_user_auditcare.py
Comment (dannyroberts) [inline corehq/apps/auditcare/management/commands/gdpr_scrub_user_auditcare.py:18]: Not sure if I'm reading this incorrectly but it looks like this is updating only what's in SQL. If we had a GDPR request before we were done migrating auditcare, we'd have to remember to re-add the version that scubs it in couch. From what I know those requests are very infrequent, but we also don't yet have a particular timeline in mind to be done migrating. Might it make sense to keep the old code for couch and just add the new code to it for SQL as well, until the migration's done?

That's kind of related to my general question of what the plan is for accessing data in couch once the models are deleted (below).

File: corehq/apps/auditcare/models.py
Comment (dannyroberts) [inline corehq/apps/auditcare/models.py]: For other couch to sql migrations I've seen we retain a Couch model and a SQL model until all the old docs are migrated. Presumably when we migrate the data having a schema for the old data will be helpful for converting it to the new format. Do you have a different approach in mind for how we'll do that?

**Review by dannyroberts — COMMENTED** [review]
This is amazing work. It all looks good to me, with the following questions. I don't think there are any major blockers, but I'll just wait to approve until I understand better after the discussion

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py]: I was planning on writing the migration against JSON read directly from Couch since there were no `wrap()` (data migration) methods on these models. I think it will be faster and pretty easy given the simplicity of these models. If they turn out to be useful, the old models can easily be revived from git history.

File: corehq/util/models.py
Comment (millerdev) [reply corehq/util/models.py]: 05f16bf6c3b52a70c14b9800539aa0af83a0e5fd

File: corehq/apps/auditcare/management/commands/gdpr_scrub_user_auditcare.py
Comment (millerdev) [reply corehq/apps/auditcare/management/commands/gdpr_scrub_user_auditcare.py]: 87ac2df454b4bcf0ebd623dbd208f014eba546b0

File: corehq/apps/auditcare/models.py
Comment (dannyroberts) [reply corehq/apps/auditcare/models.py]: OK cool, thanks for spelling that out for me, since it's a little different form what we often do. The reasoning makes sense, and I don't see any issues with it, and looking over the code I can see why not having to keep both models around makes the change simpler.

**Review by dannyroberts — APPROVED** [review]
OK all my questions are addressed. Looks like a great change. Approving, pending the steps you laid out of course.

File: corehq/apps/auditcare/tests/test_models.py
Comment (mjriley) [inline corehq/apps/auditcare/tests/test_models.py:75]: Nit: I think the test is more effective without the logic here. I'd rather see a very simple traceId spelled out in the tests. The assert is unnecessary, as if the header _isn't_ being set, then the assertion on `event.trace_id` should fail.

File: corehq/apps/auditcare/tests/test_models.py
Comment (mjriley) [inline corehq/apps/auditcare/tests/test_models.py:160]: I'd prefer that this decorator handle the case where a `trace_id_header` is already set, rather than immediately failing -- I'm sure there are cases where a test environment would want to set that value outside of these tests.

File: corehq/util/models.py
Comment (mjriley) [inline corehq/util/models.py:223]: Nit: `NullJsonField` suggests that the field stores `Null`. `DateField` stores a date, `CharField` stores characters, etc. `NullableJsonField` would be more appropriate.

**Review by mjriley — APPROVED** [review]
Looks great. Have some concerns over some of the optimizations here (the indices and foreign keys), but overall should be a huge positive to get rid of so much legacy code and move to SQL.

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py:58]: I don't think we care about having only unique values as long as duplicates don't occur very often. For example, there is a race condition in `get_or_create()` running on multiple hosts, and it's quite possible that more than one host could create a new record with the same value. Error handling for unique constraint violations seems like it would add extra complexity. I think we'd need a custom implementation of `get_or_create`.

As for making `value` the primary key, we'd still need an integer ID field since the whole point of storing the values in a separate table is to have an integer reference for long values. In addition to the issue already mentioned with unique constraints, we'd need a custom implementation of the lookup code since we could no longer use `ForeignKey` to maintain the reference.

> we're currently ignoring the primary keys entirely

They're used when reading auditcare data, which admittedly is a rare use case.

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py]: Good point. I'll remove it in a future PR.

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py]: Sorry for leaving that in there for you to review. This was eventually changed to `request.META.get("QUERY_STRING", "")`

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py]: I think it's mainly for legacy reasons. I'm also not familiar with `GenericIPAddressField`, so don't know the constraints. Is that backed by a different data type? Not sure if the historical data we have in this field in Couch would fit into it? Hopefully it would, but you never know until you've looked at the data.

I'm not sure it matters that much, to be honest. Is this a blocker for you?

File: corehq/apps/auditcare/utils/export.py
Comment (millerdev) [reply corehq/apps/auditcare/utils/export.py:62]: I'm not sure it's that important. I could just drop the assert (although don't think it hurts anything). AFAIK, asserts are always enabled in our code, even in production, and I think that's a good thing.

File: corehq/apps/auditcare/utils/export.py
Comment (millerdev) [reply corehq/apps/auditcare/utils/export.py:90]: This was copied from pre-existing code, and I didn't think about that too hard since I think this works.

I think `event_date__lte` will not work since this is coercing `end_date` to a `date` (something like `2020-03-23`) and `event_date` is a timestamp (includes a time). My understanding is that when SQL evaluates `end_date <= '2020-03-23'::datetime` (or whatever it would do in that case), it coerces `'2020-03-23'` to `'2020-03-23 00:00:00.000'`, so it would get the wrong answer for `end_date` values like `2020-03-23 12:34:56.789`.

Could revisit this in a follow-up PR, although there are other things that seem more valuable to work on since I think this code works as is.

File: corehq/util/models.py
Comment (snopoke) [reply corehq/util/models.py:314]: This definitely seems very complex to me (the whole `ForeignValue` setup). The caching in particular is very hard to reason about since we're mixing an LRU cache with a cached property (does it even live beyond the individual objects lifespan?).

I feel like a simple utility function or two would suffice and then adding methods to the model classes that use the utility functions.

**Review by snopoke — COMMENTED** [review]
I didn't do a thorough review.

File: corehq/util/models.py
Comment (millerdev) [reply corehq/util/models.py:314]: You only need this decorator if you want to pass foreign values as `__init__` arguments. I do understand the argument for using a base class instead of a decorator, although that limits the type hierarchy in other ways that doesn't seem necessary for this functionality. Implementing it as a mixin seems less than ideal because the ordering of the mixin with other base classes is very important and easy to get wrong.

Yes, a custom foreign key field implementation could probably avoid this, although that seemed more complicated.

I'm inclined to leave it for now and possibly refactor it in a follow-up PR, but probably only if it gets used more widely.

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py]: I refactored `audit.path_info` to `audit.path` and moved it to the base class (probably in a later commit than the one you're commenting on). I did not change the `request.META` field we're using since that seemed like an out of scope refactor. Could do it in a follow-up PR if you think it's important.

Out of curiosity, what is the difference between `request.META.get('PATH_INFO', '')` and `request.META.get('PATH', '')`? Is `PATH_INFO` deprecated or something?

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py:86]: My understanding is that adding the date to the index makes it much more performant to query for a user or domain in a date range, which I believe is the primary use case for queries on auditcare data. Otherwise the query planner will evaluate any date criteria as a `FILTER` condition, which gets very slow if there is a lot of data. I do realize this adds write performance overhead, but I don't think we have a choice if we want to keep queries performant. Granted, I have not measured the performance, so all this is just theorizing at this point.

My mandate in this effort has been to move auditcare from Couch to SQL, and I have tried to preserve (or improve in the case of the domain index) existing functionality as much as possible unless it is unused. Maybe it would be worth doing a separate performance analysis?

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py:224]: No. if `domain2` is `None` it can be ignored.

File: corehq/apps/auditcare/__init__.py
Comment (millerdev) [reply corehq/apps/auditcare/__init__.py:34]: `!r` uses `repr()` rather than `str()`, which makes it easier to know the type of the thing being represented.

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py:169]: Null means "unknown", empty string means the string had no value. I would prefer every field to be `NOT NULL`, but that would require a much more significant refactor and make this PR much bigger. This specific field might not need to be null, but then it would probably need to have `default=""` and all places where it is set would need to have `None` converted to empty string, which seemed to add more complexity than it was worth.

File: corehq/apps/auditcare/tests/test_models.py
Comment (millerdev) [reply corehq/apps/auditcare/tests/test_models.py:128]: The config argument is represented as part of the test name, so in the case of a test failure it is easy to see the expected value as well as the inputs.
```
FAIL: corehq.apps.auditcare.tests.test_models:test_get_domain ((expect='block', path='/a/block/path'),)
```

To me, it's a succinct way of testing multiple combinations of arguments. It was written in the spirit of the best practice outlined in _Software Engineering at Google_: "require the test author to specify only values they care about, and setting reasonable defaults for all other values."

File: corehq/util/models.py
Comment (millerdev) [reply corehq/util/models.py:223]: I don't disagree, but his code was moved from another place. Renaming the class feels out of scope for this PR.

File: corehq/apps/auditcare/migrations/0001_sqlmodels.py
Comment (millerdev) [reply corehq/apps/auditcare/migrations/0001_sqlmodels.py:52]: I think it allows us to connect a series of requests in the same session. It is a historical artifact of auditcare that we are collecting it, and removing it felt out of scope for this refactor, although I could be persuaded otherwise.

File: corehq/util/models.py
Comment (millerdev) [reply corehq/util/models.py:314]: > This definitely seems very complex to me (the whole ForeignValue setup).

To me it doesn't seem much more complicated than what's happening inside of Django's `ForeignKey` field. I tried to encapsulate the logic within a property that hopefully has clear semantics for an admittedly niche use case. Ultimately it's an optimization for a few large values associated with in a high-volume table. I think all important functionality is tested, so I think it's pretty solid.

> I feel like a simple utility function or two would suffice and then adding methods to the model classes that use the utility functions.

I didn't see another way to accomplish the same thing as succinctly without duplicating a lot of the logic on every class that uses `ForeignValue`. Can you elaborate on the simple utility functions that you're thinking of?

> mixing an LRU cache with a cached property (does it even live beyond the individual objects lifespan?)

Yes. The cache is stored on the propery instance, which is associated with the class object. It uses `cached_property` so the LRU cache is not re-created each time the property is accessed. The cache lives for the duration of the interpreter lifetime from when the property is first accessed on an instance object. That part is more complicated than I wanted, but I tried to cover it with tests to ensure it works as expected.

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py]: I'll try switching to `indexes` in a separate PR. Hopefully it does not require a database migration.

> Is the normal usage pattern to query by both user and date?

Yes. https://github.com/dimagi/commcare-hq/blob/87ac2df454b4bcf0ebd623dbd208f014eba546b0/corehq/apps/auditcare/utils/export.py#L18-L19

> If we have a composite index like this, and query only by user, do we get the same speed benefit?

Yes. However, the opposite is not true, which is why we also have an index on `event_date` alone.

File: corehq/apps/auditcare/models.py
Comment (mjriley) [reply corehq/apps/auditcare/models.py:58]: I'll leave it to your judgment, but I think ignoring the unique constraint introduces more complexity than simply handling the error. That race-condition error seems straight-forward to handle -- `get_or_create` and if we catch the error code that indicates a duplicate record was found, we just do another query and return the duplicate record.  That can be nicely encapsulated behind a function. Conversely, not handling it uniquely adds another burden of knowledge to our codebase. It would be punishing users who (correctly) prefer `get` over `filter`, thinking the tables are structured in the way we'd expect.

Agree with the point about not using them as a primary key, however.

File: corehq/apps/auditcare/models.py
Comment (mjriley) [reply corehq/apps/auditcare/models.py]: Eck...it's not a blocker, but _hqwebapp/models:UserAccessLog_ is using `GenericIPAddressField` and I'd like to be consistent if possible. Pulled up the migration code for that, and it doesn't look like its backed by a string value: `"ip" inet NULL` is the field definition. Should we just overlook the discrepancy? I'm not sure changing `UserAccessLog` would be that easy either

File: corehq/util/models.py
Comment (mjriley) [reply corehq/util/models.py:314]: Again, I'll defer to your judgment. I appreciate that this lets me use these models in a more efficient fashion, but it still feels pretty cute...

File: corehq/apps/auditcare/models.py
Comment (mjriley) [reply corehq/apps/auditcare/models.py]: From the [docs](https://docs.djangoproject.com/en/3.1/ref/request-response/#django.http.HttpRequest.path_info): `For example, if the WSGIScriptAlias for your application is set to "/minfo", then path might be "/minfo/music/bands/the_beatles/" and path_info would be "/music/bands/the_beatles/"`. I'm guessing the distinction probably doesn't matter for our application? But neither are deprecated, and they do mean different things.

File: corehq/apps/auditcare/migrations/0001_sqlmodels.py
Comment (mjriley) [reply corehq/apps/auditcare/migrations/0001_sqlmodels.py:89]: It really only hurts manually querying the SQL. If you think that's a pretty rare edge-case, then this makes sense. If we reasonably will be interacting with the raw SQL, I'd rather leave it with the verbose names

File: corehq/apps/auditcare/models.py
Comment (mjriley) [reply corehq/apps/auditcare/models.py:86]: Possibly, but I think this might be more my unfamiliarity with multi-indices. From [Postgres docs](https://www.postgresql.org/docs/10/indexes-multicolumn.html), it seems like these will be very efficient for user or domain queries. It seems reasonable that we might want to retrieve all logs for a given date range, and then I'm not sure how performant these indices would be, but I'm just spitballing.

File: corehq/apps/auditcare/tests/test_models.py
Comment (mjriley) [reply corehq/apps/auditcare/tests/test_models.py:128]: Sorry, indulge me for a wall-of-quotes from the same chapter.


>When a test fails, an engineer’s first job is to identify which of these cases the failure falls into and then to diagnose the actual problem. The speed at which the engineer can do so depends on the test’s clarity. A clear test is one whose purpose for existing and reason for failing is immediately
clear to the engineer diagnosing a failure. Tests fail to achieve clarity when their reasons for failure aren’t obvious or when it’s difficult to figure out why they were originally written. Clear tests also bring other benefits, such as documenting the system under test and more easily serving as a basis for new tests.

>For a test suite to scale and be useful over time, it’s important that each individual test in that suite be as clear as possible.

If any of these tests failed, the intent of each test is not immediately apparent.

>A test is complete when its body contains all of the information a reader needs in order to understand how it arrives at its result. A test is concise when it contains no other distracting or irrelevant information... a test's body should contain all the information needed to understand it without containing any irrelevant or distracting information

This is the basis for the quote you cited -- avoiding extraneous information. This helper does the opposite -- it obscures the expectation. If we needed to provide a port to make our request, that would be a great thing to hide in a helper method, because it's irrelevant to what we're testing; conversely, the expectation is critical to what we're testing.

>Remember that every behavior has three parts: a "given" component that defines how the system is set up, a "when" component that defines the action to be taken on the system, and a "then" component that validates the result. Tests are clearest when this structure is explicit.

>The test name is very important: it will often be the first or only token visible in failure reports, so it's your best opportunity to communicate the problem when the test breaks. It's also the most straightforward way to express the intent of the test.

>A good name describes both the actions that are being taken on a system and the expected outcome.

>Names like this are much more verbose than we’d normally want to write for methods in production code, but the use case is different: we never need to write code that calls these, and their names frequently need to be read by humans in reports. Hence, the extra verbosity is warranted.

Parameterized tests like these lose all of this information, and thus become less clear. Here's my understanding of what each is testing (this wasn't immediately apparent to me):
- When a path contains no domain, None should be returned: `path_lacking_domain_returns_None`
- When a path contains a domain, that domain should be returned: `path_including_domain_returns_domain`
- When a request contains both a path and an explicit domain, the explicit domain is used if the path does not contain one: `path_lacking_domain_returns_domain_param_when_domain_param_exists`
- When a request contains both a path with a domain and an explicit domain, the path's domain is used: `path_including_domain_is_used_over_domain_param`

>in test code, stick to straight-line code over clever logic, and consider tolerating some duplication when it makes the test more descriptive and meaningful.

>Clear tests are trivially correct upon inspection.

>These tests have more duplication, and the test bodies are a bit longer, but the extra verbosity is worth it. Each individual test is far more meaningful and can be understood entirely without leaving the test body. A reader of these tests can feel confident that the tests do what they claim to do and aren’t hiding any bugs.

Verbosity isn't a bad thing with tests -- logic in code generally is. These tests, ideally, will be written once and won't change.

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py]: Interesting. It shouldn't matter for our application since I think `PATH` and `PATH_INFO` will always be the same (we do not set `WSGIScriptAlias`).

File: corehq/apps/auditcare/tests/test_models.py
Comment (millerdev) [reply corehq/apps/auditcare/tests/test_models.py:160]: `assert_not_installed` is used to verify test pre- and post-conditions. Pre-conditions are asserted to confirm that the tests will not produce a false positive. Post-conditions are asserted to verify that cleanup was successful.

I suppose the pre-conditions could be made a bit more flexible with
```py
assert AccessAudit.trace_id_header != settings.AUDIT_TRACE_ID_HEADER, ...
```
I could also set `TRACE_HEADER` to a value that is less likely to collide with values that someone might use in localsettings. Would that be better?

File: corehq/util/models.py
Comment (snopoke) [reply corehq/util/models.py:314]: > > This definitely seems very complex to me (the whole ForeignValue setup).
> 
> To me it doesn't seem much more complicated than what's happening inside of Django's `ForeignKey` field. I tried to encapsulate the logic within a property that hopefully has clear semantics for an admittedly niche use case. Ultimately it's an optimization for a few large values associated with in a high-volume table. I think all important functionality is tested, so I think it's pretty solid.

For sure it's not more complex than that but it's a lot of complexity for a very niche use case and it does make it hard to reason about what's going on. The Django FK models are well known and understood and match what's happening at the DB layer, this is not. 
 
> > I feel like a simple utility function or two would suffice and then adding methods to the model classes that use the utility functions.
> 
> I didn't see another way to accomplish the same thing as succinctly without duplicating a lot of the logic on every class that uses `ForeignValue`. Can you elaborate on the simple utility functions that you're thinking of?

Sure there would be some duplication but since this is essentially just an abstraction on a lookup you could do something like this:
```python
def get_user_agent(self):
    return get_value_from_foreign_key(UserAgent, self.user_agent_fk_id)

def set_user_agent(self, value):
   return self.user_agent_fk = get_or_create_value(UserAgent, value)
```

> > mixing an LRU cache with a cached property (does it even live beyond the individual objects lifespan?)
> 
> Yes. The cache is stored on the propery instance, which is associated with the class object. It uses `cached_property` so the LRU cache is not re-created each time the property is accessed. The cache lives for the duration of the interpreter lifetime from when the property is first accessed on an instance object. That part is more complicated than I wanted, but I tried to cover it with tests to ensure it works as expected.

Ah, I had it in my head that this was on the model instance but it's on they `ForeignValue` instance which has a different lifecycle to the model. Still, I think that just illustrates how hard this can be to reason about.

Considering that ALL usages of this make use of the cache I would recommend simplifying the code to make caching mandatory instead of optional.

**Review by snopoke — COMMENTED** [review]
I'm surprised this got merged so quickly. It seems like there was still a lot of discussion going on. Also considering this is a high risk area the PR should be labelled as such and I would have expected some QA on staging.

File: corehq/apps/auditcare/tests/test_models.py
Comment (mjriley) [reply corehq/apps/auditcare/tests/test_models.py:160]: Yep!

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py:58]: This turned out to be much worse than I had anticipated. Fixes in https://github.com/dimagi/commcare-hq/pull/29391 and https://github.com/dimagi/commcare-hq/pull/29393

File: corehq/util/models.py
Comment (millerdev) [reply corehq/util/models.py:314]: Upon revisiting this after the initial rollout drama, I'm going to keep the `cache_size` parameter since it was useful to be able to disable the cache temporarily during a database migration.

While I concede that there is a fair amount going on in `ForeignValue`, I still think it is a useful abstraction neatly encapsulated within a property descriptor. I think it is significant performance optimization for auditcare, saving up to two database round-trips per request.

It may make sense to limit its availability scope by moving it into the auditcare app if/when `UserAccessLog` gets merged into auditcare.

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py]: https://github.com/dimagi/commcare-hq/pull/29457/commits/f90846400144e1f2fc05867f89a5909c59744d5b

File: corehq/apps/auditcare/models.py
Comment (millerdev) [reply corehq/apps/auditcare/models.py]: https://github.com/dimagi/commcare-hq/pull/29457/commits/0818bfd14f49f2034e1346e3c39e2b7756e7f719

File: corehq/apps/auditcare/utils/export.py
Comment (millerdev) [reply corehq/apps/auditcare/utils/export.py:85]: https://github.com/dimagi/commcare-hq/pull/29457/commits/d257d1b85892957072626b3bc76f67b1ee362e42

File: corehq/apps/auditcare/tests/test_models.py
Comment (millerdev) [reply corehq/apps/auditcare/tests/test_models.py:160]: https://github.com/dimagi/commcare-hq/pull/29457/commits/1e5e206a66bd397f8605d902628eec95bf031c0b

**Pull request merged**
